src/RecyclableMemoryStream.cs

```diff
@@ -626,10 +620,33 @@ public override string ToString()
         /// <exception cref="ObjectDisposedException">Object has been disposed</exception>
         public override void WriteByte(byte value)
         {
-            this.CheckDisposed();
-            this.byteBuffer[0] = value;
-            this.Write(this.byteBuffer, 0, 1);
-        }
+			this.CheckDisposed();
+			int blockSize = this.memoryManager.BlockSize;
+			int end = this.position + 1;
+			// Check for overflow
+			if (end > MaxStreamLength)
+			{
+				throw new IOException("Maximum capacity exceeded");
+			}
+
+			this.EnsureCapacity(end);
+
+			if (this.largeBuffer is null)
+			{
+				int block = this.position / blockSize, offset = this.position % blockSize;
+
```

Comment:
`offset = this.position - block * blockSize` will save "expensive" modulo operation, you already have the division result.